### PR TITLE
Use docker/metadata-action for deploy-DEV.yml

### PR DESCRIPTION
**Pull request by echoix**
Since the workflow of DEV-linters worked correctly in the last couple of weeks, including with dependabot PRs, I applied a similar patch for the DEV workflow. This will allow all checks to be run for dependabot PRs. At the moment, these changes do not touch images that would be pushed to registry. 

This is a reduced version of my experiments of yesterday, where I managed to add more advanced metadata to the manifests, had OCI manifest images instead of docker manifest images while still having them used in the same workflow (ie: what was blocking us of building and loading multiplatform images @bdovaz !!), and even was able to use some advanced exporters, like eStargz and zstd. (Only zstd managed to be a little faster at the end of the build, but had to pull later on.) However I’m not finished exploring it since I hit some no more space after the tests, probably since I had two copies of the image. I was using a service container for the job, with a local registry (at localhost:5000), that we can push to without it being on docker hub or other. 

Some nice things might be close ;)

## Proposed Changes

1. …
2. …
3. …

### Author/Contributor
- [ ] Add entry to the [CHANGELOG](https://github.com/oxsecurity/megalinter/blob/main/CHANGELOG.md) listing the change and linking to the corresponding issue (if appropriate)
- [ ] If documentation is needed for this change, has that been included in this pull request

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [x] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

### CODE REVIEW

**Comment by nvuillam** [comment]
@echoix as I'm a docker fraud (like on many on the topics) I don't understand in details everything you are saying, but it seems convincing, so if it ensures ascending compatibility to current MegaLinter users, I'm all in :)

For a v8, usually it requires time, if/when we succeed to have multiplatform builds, I'm more than ok to take some time to release it :)

**Comment by echoix** [comment]
This PR does nothing to change externally (that's why I didn't apply yet in this step to beta and release workflows). And also why I didn't wait to have completely finished with my experiments, at least having the CI work for other auto updates means I'm able to maintain it during the week, if it only requires analyzing and merging. Coding takes a bit more time.

**Comment by echoix** [comment]
I think we could easily progressively switch all of our workflows to metadata actions while in v7, and then add the changes that make it a little bit more different in v8. This means to not use `sbom: true`, `provenance: true`, `provenance: mode=max` or `annotations: ${{ steps.meta.outputs.annotations }} ` in the docker/build-push-action (as it requires creates a manifest index, instead of just one image, the same that enables to have multiplatform images).

**Comment by echoix** [comment]
Converting to draft since the same actions ran in my fork had an unexpected failure and I want to take a look before

File: .github/workflows/deploy-DEV.yml [hunk 1]
```diff
@@ -259,7 +234,7 @@ jobs:
         run: cd mega-linter-runner && sudo yarn install --frozen-lockfile && sudo npm link
       - name: Run mega-linter-runner tests
         if: ${{ steps.docker_build.outcome }} == 'success' && !contains(github.event.head_commit.message, 'quick build')
-        run: cd mega-linter-runner && MEGALINTER_RELEASE=${{ steps.image_tag.outputs.tag }} MEGALINTER_NO_DOCKER_PULL=true npm run test
+        run: cd mega-linter-runner && MEGALINTER_RELEASE=${{ fromJSON(steps.meta.outputs.json).labels['org.opencontainers.image.version'] }} MEGALINTER_NO_DOCKER_PULL=true npm run test
```
[the remark this answers is not in the document]
Comment (echoix) [reply .github/workflows/deploy-DEV.yml:237]: Fixing this and the PR would be good to go (with a changelog entry) would there be better ways to not having to go back and manually write a changelog entry once we know the PR number? It's the single thing that chills me off of doing a PR to fix something.

**Comment by echoix** [comment]
Ok, what I’d want:
Something similar to: https://github.com/oxsecurity/megalinter/blob/d85dff0ecd96a54f017cb310aefb6af3f46295f3/mega-linter-runner/test/megalinter-cli.test.js#L7


But for the `image` option. https://github.com/oxsecurity/megalinter/blob/d85dff0ecd96a54f017cb310aefb6af3f46295f3/mega-linter-runner/lib/options.js#L41-L51


Is it possible to pass other args to the command without changing the test code? Or is it really needed to add a env var flag for that?

**Comment by echoix** [comment]
I think I managed to do something that could work. Please wait for me to check that the same workflows pass in my repo too to make sure that it works properly (as it should work here, but not necessarily elsewhere)

**Comment by echoix** [comment]
It « works » as is doesn’t fail, but if you look at this step (of the mega linter runner tests (npm run test)), the at the very top, it pulls the docker image for oxsecurity/megalinter:beta, but I think I added the `--image` argument everywhere I think it should have been used. 

https://github.com/echoix/megalinter/actions/runs/7217127730/job/19664442064#step:18:33

Did I miss something? Which test step calls that? I hope it doesn’t need to pull the image only to display help.

And I think I realized that maybe that was the issue that had my experiment of using a local registry end up with no space left: It might had pulled an other image at this step, thus having 2 (or more) copies of Megalinter on disk, running out of space

**Comment by nvuillam** [comment]
I see you use MEGALINTER_IMAGE env var, but when is it supposed to be set in the workflows ?

File: .github/workflows/deploy-DEV.yml [hunk 2]
```diff
@@ -259,7 +234,7 @@ jobs:
         run: cd mega-linter-runner && sudo yarn install --frozen-lockfile && sudo npm link
       - name: Run mega-linter-runner tests
         if: ${{ steps.docker_build.outcome }} == 'success' && !contains(github.event.head_commit.message, 'quick build')
-        run: cd mega-linter-runner && MEGALINTER_RELEASE=${{ steps.image_tag.outputs.tag }} MEGALINTER_NO_DOCKER_PULL=true npm run test
+        run: cd mega-linter-runner && MEGALINTER_RELEASE=${{ fromJSON(steps.meta.outputs.json).labels['org.opencontainers.image.version'] }} MEGALINTER_IMAGE=${{ fromJson(steps.meta.outputs.json).tags[0]}} MEGALINTER_NO_DOCKER_PULL=true npm run test
```
Comment (echoix) [inline .github/workflows/deploy-DEV.yml:237]: The MEGALINTER_IMAGE is set here, and only used for the test scripts. Internal only. It’s the way I can add an argument to the options

Comment (nvuillam) [reply .github/workflows/deploy-DEV.yml:237]: ok :)

**Comment by echoix** [comment]
What I'm not sure, is are the other tests above run before, and do they need to pull the image in order to display help, or upgrade the config files? It shouldn't

**Comment by nvuillam** [comment]
@echoix so it is ok to merge ? if u are around maybe u can update the changelog.md ?  :)

**Comment by echoix** [comment]
Well, I still need help to understand why a pull of the beta image, not the one just build, is done for some of the tests in npm test, as it is unexpected. And will need to be fixed before I can do another step with this line of changes.

**Comment by echoix** [comment]
I'm actively debugging this this morning

**Comment by echoix** [comment]
I think I found something: 

image

In the mocha test thingy, after the options is parsed, it seems to have nothing that we want and the mocha arguments instead.

Let's start over. I added breakpoints almost everywhere. I start debugging. The megalinter-cli.test.js file is passed over by only checking the it() calls, looks like it adds them to a list of tests that will be executed, tests aren't run yet. Then some work is done, and now the actual running of our tests happen. 

One of the first breakpoints that hit, if I go just under in the call stack, we see that the MegaLinterRunnerCli module/class (I don't know what they call it in node) is instantiated with process.argv. Seems ok, but in tests, it isn't the good ones.

image
These are the test ones, not for the runner.
The next two images are for the same breakpoint:
image
image

Then, the parsing of options happen, going through the optionator code.

image



The rest will follow...

**Comment by echoix** [comment]
Please wait if this is enough to make it work properly. If so, I'll make a real PR to have it published as a beta npm package, and do a last test (when not inside a Megalinter repo), and simply "installing" that version of the runner, and make sure that user's side still work correctly.

**Comment by echoix** [comment]
Once https://github.com/oxsecurity/megalinter/pull/3233 is merged, this PR could be updated (or not), and directly be merged. Its ready.

**Pull request merged**